nemo_aligner/data/nlp/samplers.py

```diff
@@ -0,0 +1,98 @@
+# Copyright (c) 2023, NVIDIA CORPORATION. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from itertools import chain
+from typing import Optional
+
+import torch
+
+from nemo.collections.nlp.data.language_modeling.megatron.data_samplers import BaseMegatronSampler
+
+from nemo.utils import logging
+
+
+class MegatronPretrainingRandomSampler(BaseMegatronSampler):
```

Comment:
It supports multi-epoch training. But as I had mentioned elsewhere we should eventually upstream this change into NeMo.

Reply:
can we upstream this before the PR? I don't think it'll take long but I'm hoping to not hold code like this in aligner(since now someone has to remember to delete it)

Reply:
> can we upstream this before the PR? I don't think it'll take long but I'm hoping to not hold code like this in aligner(since now someone has to remember to delete it)

I understand your point but unless we sync our main branch with NeMo's main (which we decided not to do) then I'm afraid this is how we'll need to do things (implement stuff in Aligner then upstream it afterwards).

Reply:
ah ok then for things we plan to upstream can we at least file the PR to NeMo and link it here to flag for removal? just so we don't lose track

Reply:
Remember how long it took to get the Nemo upstream PR for RandomBatchSampler accepted? Several months, so it makes sense to put stuff in Aligner to start with. Also, aren't we planning to decouple from Nemo entirely? So then won't we need to have these classes in Aligner anyway?

Reply:
How do you want me to mark it?

Reply:
file a PR to nemo with this sampler, and link it here. Make a note saying once that PR is merged we should delete this class.

Reply:
Make a note in the source code of the sampler class here in Aligner or a note in this PR?

Reply:
Here is the upstream PR for Nemo: https://github.com/NVIDIA/NeMo/pull/8239

Reply:
> Make a note in the source code of the sampler class here in Aligner or a note in this PR?

Should be in the code too, will be easier to track it down later.

Reply:
Done